Cluster recovery (#340)
canonical/microcloud: doc/how-to/recover.md

> # MicroCloud Cluster Recovery

Reviewer: ```suggestion
# How to recover a MicroCloud cluster
```

> Each MicroCloud service uses the [Dqlite](https://dqlite.io/) distributed

Reviewer: You'll need to add Dqlite to the wordlist here as well - there's no automatic update (unfortunately :disappointed: ).

Author: K, wasn't sure if there was a process for pulling in starter pack changes.

database for highly-available storage. While the cluster recovery process is
similar for each service, this document only covers cluster recovery for the

> `microcloudd` daemon. For cluster recovery procedures for LXD, MicroCeph and

Reviewer: Just to confirm - is two 'd' correct?

Author: [Yeah](https://github.com/canonical/microcloud-pkg-snap/blob/latest-edge/snapcraft/commands/daemon.start#L4), my intention was to refer to just the microcloud daemon (to avoid ambiguity with the MicroCloud project/product).

Reviewer 2: From an external user's point of view, is `microcloudd` ever actually exposed? Would they know what this is referring to?

Author: `snap services` (`microcloud.daemon`), or `ps -aux`; IMO an operator should be familiar with what is in scope/responsibilities for each tool (especially since almost all use of a deployed MicroCloud is done via LXD/Micro{Ceph,OVN}). I think `d` naming convention + context here makes the meaning clear.
You're welcome to disagree :)

MicroOVN, see:

- [LXD Cluster Recovery](https://documentation.ubuntu.com/lxd/en/latest/howto/cluster_recover/)
- [MicroOVN Launchpad Bug](https://bugs.launchpad.net/microovn/+bug/2072377)
- [MicroCeph Issue](https://github.com/canonical/microceph/issues/380)

> If a MicroCloud cluster loses more than a quorum of its database members, then

Reviewer 2: Would be nice to define what is the minimum qualification of "quorum"

Author: I was initially going to go into way too much detail here regarding quorum and member roles, but that stuff is really more in scope in the Dqlite docs, not here. That said, the dqlite docs [do not include](https://github.com/canonical/dqlite/issues/369) any information on member roles beyond what's present in the [wire protocol spec](https://dqlite.io/docs/reference/wire-protocol).
I rearranged this a bit and I think it improves clarity.

database operations will no longer be possible on the entire cluster. If the
loss of quorum is temporary (e.g. some members temporarily lose power), database

> operations will be restored when the down members come back online.
> 
> This document describes how to recover database access if the down members have

Reviewer: ```suggestion
operations will be restored when the offline members come back online.

This document describes how to recover database access if the offline members have
```

been lost without the possibility of recovery (e.g. disk failure).

> ## Recovery Procedure

Reviewer: ```suggestion
## Recovery procedure
```

> 1. In order to perform cluster recovery, **all cluster members must be shut down**:

Reviewer: ```suggestion
1. Shut down all cluster members before performing cluster recovery:
```

We don't use bold for emphasis - better to reword so the most important part comes first.

> ```
> sudo snap stop microcloud
> ```

Reviewer: ````suggestion
   ```
   sudo snap stop microcloud
   ```
````

This is a global thing - indent everything that should be under a step with three spaces.

1. Once all cluster members are shut down, determine which Dqlite database is
most up to date. Look for files in `/var/snap/microcloud/common/state/database` whose
filenames are two numbers separated by a dash (i.e.
`0000000000056436-0000000000056501`). The largest second number in the directory

> is the end-index of the most recently closed segment (56501 in the example).
> Perform the next step on the cluster member with the highest end-index.

Reviewer: ```suggestion
is the end index of the most recently closed segment (56501 in the example).
Perform the next step on the cluster member with the highest end index.
```

1. Use the following command **on one cluster member** to reconfigure the Dqlite

>    roles for each member:

Author: Sorry, missed it. I added a clarification to the preceding para as well.

> 2. Use the following command **on one cluster member** to reconfigure the Dqlite
> roles for each member:
> ```
> sudo microcloud cluster edit
> ```

Reviewer: ````suggestion
1. On the cluster member with the most up-to-date database, reconfigure the Dqlite roles for each member with the following command:
   ```
   sudo microcloud cluster edit
   ```
````

All steps should use `1.` - with the correct indentation this should work. :wink:
The advantage is that you can later insert a step and don't need to think about updating the numbers.

3. As indicated by the output of the above command, copy
`/var/snap/microcloud/common/state/recovery_db.tar.gz` to the same path on each
cluster member.

4. Restart MicroCloud. The recovered database tarball will be loaded on daemon
startup. Once a quorum of voters have been started, the MicroCloud database
will become available.
```
sudo snap start microcloud
```

## Backups

MicroCloud creates a backup of the database directory before performing the
recovery operation to ensure that no data is lost. The backup tarball is created

> in `/var/snap/microcloud/common/state/`. In case the cluster recovery operation

Reviewer: ```suggestion
in `/var/snap/microcloud/common/state/`. If the cluster recovery operation
```

# How to recover a MicroCloud cluster

```{note}
Each MicroCloud service uses the [Dqlite](https://dqlite.io/) distributed
database for highly-available storage. While the cluster recovery process is
similar for each service, this document only covers cluster recovery for the
`microcloudd` daemon. For cluster recovery procedures for LXD, MicroCeph and
MicroOVN, see:

- [LXD Cluster Recovery](https://documentation.ubuntu.com/lxd/en/latest/howto/cluster_recover/)
- [MicroOVN Launchpad Bug](https://bugs.launchpad.net/microovn/+bug/2072377)
- [MicroCeph Issue](https://github.com/canonical/microceph/issues/380)
```

> Dqlite requires a majority of its database voters (a quorum) to be accessible in

Reviewer 2: Hm, "a majority of its database voters" can sound very misleading. I believe we cap out at 4 voters at a time, so this might make someone think they need at least 3 nodes up at all times, when really it's 2.

How about this:
```
For MicroClouds with 2 or more systems, at least 2 database voters (a quorum) must be accessible in order to perform database operations
```

Reviewer 2: Hmm, I thought I recalled at least LXD hitting 4 voters, 2 standby, 1 spare but you're right, both LXD and microcluster have the max voters set to 3. 

In that case I take it back, "majority" is correct, because we may configure it someday and have to change the docs. Sorry for the mislead.
